larray/core/session.py

```diff
@@ -1392,6 +1393,155 @@ def display(k, v, is_metadata=False):
+
+    def __setattr__(self, key, value):
+        if key != 'meta':
+            self._check_key_value(key, value)
+
+        # update the real attribute
+        object.__setattr__(self, key, value)
+        # update self._objects
+        Session.__setattr__(self, key, value)
+
+    def _check_key_value(self, key, value):
+        cls = self.__class__
+        attr_def = getattr(cls, key, None)
+        if attr_def is None:
+            raise ValueError("The '{item}' item has not been found in the '{cls}' class declaration. "
+                             "Adding a new item after creating an instance of the '{cls}' class is not permitted."
+                             .format(item=key, cls=cls.__name__))
+        if (isinstance(value, (int, float, basestring, np.generic)) and value != attr_def) \
+                or (isinstance(value, (Axis, Group)) and not value.equals(attr_def)):
+            raise TypeError("The '{key}' item is of kind '{cls_name}' which cannot by modified."
+                            .format(key=key, cls_name=attr_def.__class__.__name__))
+        if type(value) != type(attr_def):
+            raise TypeError("Expected object of type '{attr_cls}'. Got object of type '{value_cls}'."
+                            .format(attr_cls=attr_def.__class__.__name__, value_cls=value.__class__.__name__))
+        if isinstance(attr_def, Array):
+            try:
+                attr_def.axes.check_compatible(value.axes)
+            except ValueError as e:
+                msg = str(e).replace("incompatible axes:", "Incompatible axes for array '{key}':".format(key=key))
+                raise ValueError(msg)
+        elif isinstance(value, np.ndarray) and value.shape != attr_def.shape:
+            raise ValueError("Incompatible shape for Numpy array '{key}'. "
+                             "Expected shape {attr_shape} but got {value_shape}."
+                             .format(key=key, attr_shape=attr_def.shape, value_shape=value.shape))
+
+    def copy(self):
+        instance = self.__class__()
+        for key, value in self.items():
+            instance[key] = copy(value)
+        return instance
+
+    def apply(self, func, *args, **kwargs):
+        kind = kwargs.pop('kind', Array)
+        instance = self.__class__()
+        for key, value in self.items():
+            instance[key] = func(value, *args, **kwargs) if isinstance(value, kind) else value
+        return instance
+
+    def _disabled(self, *args, **kwargs):
+        """This method will not work because adding or removing item and modifying axes of declared arrays
+        is not permitted."""
+        raise ValueError(
+            "Adding or removing item and modifying axes of declared arrays is not permitted.".format(
+                cls=self.__class__.__name__
+            )
+        )
+
+    # XXX: not sure we should or not disable 'transpose()'?
+    __delitem__ = __delattr__ = _disabled
+    update = filter = transpose = compact = _disabled
```

Comment:
`update()` should not be disabled because it can be used to change the values of several *existing* arrays in one shot. 

Only adding new arrays should trigger a warning (or be disallowed, depending on the point of view) but this is already taken care of by `__setitem__()`, which `update()` uses, so it is not even necessary to override `update()`.

IMO, the other 3 methods should not be disabled either since they (should) just return a new (non frozen) Session.